web/docs/auth/auth-hooks.md

```diff
@@ -0,0 +1,590 @@
+---
+title: Auth Hooks
+---
+
+import { EmailPill, UsernameAndPasswordPill, GithubPill, GooglePill, KeycloakPill } from "./Pills";
+import ImgWithCaption from '@site/blog/components/ImgWithCaption'
+
+Auth hooks allow you to "hook into" the auth process at various stages and run your custom code. For example, if you want to forbid certain emails from signing up, or if you want to send a welcome email to the user after they sign up, auth hooks are the way to go.
+
+## Using hooks
+
+To use auth hooks, you must first declare them in the Wasp file:
+
+<Tabs groupId="js-ts">
+<TabItem value="js" label="JavaScript">
+
+```wasp
+app myApp {
+  wasp: {
+    version: "^0.13.0"
+  },
+  auth: {
+    userEntity: User,
+    methods: {
+      ...
+    },
+    onBeforeSignup: import { onBeforeSignup } from "@src/auth/hooks",
+    onAfterSignup: import { onAfterSignup } from "@src/auth/hooks",
+    onBeforeOAuthRedirect: import { onBeforeOAuthRedirect } from "@src/auth/hooks",
+  },
+}
+```
+</TabItem>
+<TabItem value="ts" label="TypeScript">
+
+```wasp
+app myApp {
+  wasp: {
+    version: "^0.13.0"
+  },
+  auth: {
+    userEntity: User,
+    methods: {
+      ...
+    },
+    onBeforeSignup: import { onBeforeSignup } from "@src/auth/hooks",
+    onAfterSignup: import { onAfterSignup } from "@src/auth/hooks",
+    onBeforeOAuthRedirect: import { onBeforeOAuthRedirect } from "@src/auth/hooks",
+  },
+}
+```
+</TabItem>
+</Tabs>
+
+If the hooks are defined as async functions, Wasp _awaits_ them. This means the auth process waits for the hooks to finish before continuing.
+
+Wasp ignores the return values of the hooks. The only exception is the `onBeforeOAuthRedirect` hook, whose return value affects the OAuth redirect URL.
+
+### When are the hooks called
+
+Below you can see the flow of the auth process with hooks:
+
+<ImgWithCaption
+  source="/img/auth-hooks/signup_flow_with_hooks.png"
+  alt="Signup Flow with Hooks"
+  caption="Signup Flow with Hooks"
+/>
+
+<ImgWithCaption
+  source="/img/auth-hooks/oauth_flow_with_hooks.png"
+  alt="OAuth Flow with Hooks"
+  caption="OAuth Flow with Hooks"
+/>
+
+## Available hooks
+
+Wasp supports the following auth hooks:
+- `onBeforeSignup`
+- `onAfterSignup`
+- `onBeforeOAuthRedirect`
+
+Let's go through each of them.
+
+### Executing code before the user signs up
+
+Wasp calls the `onBeforeSignup` hook before the user is created.
+
+The `onBeforeSignup` hook can be useful if you want to reject a user based on some criteria before they sign up.
+
+Works with <EmailPill /> <UsernameAndPasswordPill /> <GithubPill /> <GooglePill /> <KeycloakPill />
```

Comment:
```suggestion
This hook works with with <EmailPill /> <UsernameAndPasswordPill /> <GithubPill /> <GooglePill /> <KeycloakPill />
```
To avoid a sentence fragment.